### ISSUE TITLE
Implement backref type

### ISSUE THREAD

**Issue by sirex** [atviriduomenys/spinta#96]
`backref` type can improve data navigation and also would add possibility to do more advanced filtering.

### One to Many

```
d | r | b | m | property    | type    | ref      | access
$DATASET                    |         |          |
                            |         |          |
  |   |   | City            |         |          |
  |   |   |   | name        | string  |          | open
  |   |   |   | country     | ref     | Country  | open
                            |         |          |
  |   |   | Country         |         |          |
  |   |   |   | name        | string  |          | open
  |   |   |   | cities[]    | backref | City     | open
```

Here we have **one** Country and **many** City objects.

`Country.cities[]` is a `backref`, that means, we can do queries like this:

    /Country/cities

And this gives us list of all `City` objects, that have a country.

Also we can do things like this:

    /Country/a25e183c-e31d-4ad8-ad9a-48b91b810783/cities

This will return list of all cities in a country.

By default, when querying:

    /Country

data of `backref` type properties are not returned giving only empty lists.
Here is what we will get with `/Country` query:

```json
{
    "_data": [
        {
            "_id": "a25e183c-e31d-4ad8-ad9a-48b91b810783",
            "name": "Lithuania",
            "cities": []
        }
    ]
}
```

### expand() function

In order, to also include data of `backref` properties, we need to use `expand()` function.

We can automatically include all `backref` data, by calling `expand()`:

    /Country?expand()

This will include all data, from all the `backref` columns, and we will get
following response:

```json
{
    "_data": [
        {
            "_id": "a25e183c-e31d-4ad8-ad9a-48b91b810783",
            "name": "Lithuania",
            "cities": [
                {"_id": "0d430d92-574e-45da-8cd2-12a05e73c2b6"},
                {"_id": "c2c97660-b5e8-4f08-850f-c9ebe46e3798"},
                {"_id": "fee66313-4da6-4469-a6bb-b993fe11bd19"}
            ]
        }
    ]
}
```

`expand()` function, accepts `*args`, to specify `backref` property names to
expand. For example:

    /Country?expand(cities)

Only `cities` are expanded.

### Many to Many

```
d | r | b | m | property    | type    | ref      | access
$DATASET                    |         |          |
                            |         |          |
  |   |   | Language        |         |          |
  |   |   |   | name        | string  |          | open
  |   |   |   | countries[] | backref | Country  | open
                            |         |          |
  |   |   | Country         |         |          |
  |   |   |   | name        | string  |          | open
  |   |   |   | languages[] | ref     | Language | open
```

Here we have **many** `Country` and **many** `Language` objects.

Everything works exactly the same as with One to Many relation, except here
`Country.languages` is an `array` type. See

- https://github.com/atviriduomenys/spinta/issues/161

### One to One

`backerf` can also be used for one to one relations, like here:

```
d | r | b | m | property    | type    | ref      | access
$DATASET                    |         |          |
                            |         |          |
  |   |   | Country         |         |          |
  |   |   |   | name        | string  |          | open
  |   |   |   | capital     | ref     | Capital  | open
                            |         |          |
  |   |   | Capital         |         |          |
  |   |   |   | name        | string  |          | open
  |   |   |   | country     | backref | Country  | open
```

Here we can access

    /Country?select(name, capital.name)

and

    /Capital?select(name, country.name)

both ways.

### level < 4

When `property.level` is less than 4, then we follow rules as described here:

- https://atviriduomenys.readthedocs.io/api/index.html#ir-zemesnis-brandos-lygis

Basically, instead of using `_id` as an identifier, we use `model.ref` as
identifier.

### Relation with array type

Properties of `array` type, store data in a separate table, but when `array`
items are of `backref` type, then we don't need a separate table.

### Related issues

- [x] https://github.com/atviriduomenys/spinta/issues/161

### Tasks

- [ ] https://github.com/atviriduomenys/spinta/issues/512 - Add URI support for backrefs
- [ ] Add support for RDF format, https://github.com/atviriduomenys/spinta/blob/10717c9c57cc95992a44523ff0f57f3cb053664d/notes/types/backref.sh#L249-L265

**Comment by sirex** [comment]
https://github.com/atviriduomenys/spinta/pull/515 pull request can be merged, but I moved this task back to Ready, because important RDF support is still missing.

**Comment by JustinasKen** [comment]
Currently `Backref` only supports basic access, if there is a need to be able to support `Denorm` properties, need to make another task and rework part of the logic.
